ENH: Support ExtensionArray operators via a mixin

Dr-Irv opened the pull request:
- [x] closes #19577
- [x] tests added / passed
    - tests/extension/decimal/test_decimal.py:TestOperator
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry
    - in v0.24.0

Based on a discussion in #20889, this provides a mixin (via a mixin factory) that provides a default implementation of operators for `ExtensionArray` using the operators defined on the underlying `ExtensionDtype` .  Tested using the `DecimalArray` implementation.

NOTE: This requires #21183 ~~and #21260~~ to be accepted into master, and so the changes from those pull requests are included here.

Comments from @jorisvandenbossche @TomAugspurger @jreback @jbrockmendel are welcome.

Dr-Irv committed:
ENH: Support ExtensionArray operators via a mixin

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +615,91 @@ def _ndarray_values(self):
+                    ovalues = [parm] * len(self)
+                return ovalues
+            lvalues = convert_values(self)
+            rvalues = convert_values(other)
```
This should probably do alignment as well?

Dr-Irv replied:
It doesn't need to, for a few reasons:
  - The `_binop` method from here is a method on `ExtensionArray` (i.e., `isinstance(self, ExtensionArray) == True`
  - This method is called by `dispatch_to_extension_ops` in `ops.py`, which is called after alignment has already been done in `_arith_method_SERIES` and alignment is ensured in `_comp_method_SERIES`

jorisvandenbossche replied:
Ah, OK, this is also tested?

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +615,91 @@ def _ndarray_values(self):
+                    return f(b)
+                else:
+                    return NotImplemented
+            res = [callfunc(a, b) for (a, b) in zip(lvalues, rvalues)]
```
If we would do `op(a, b)` directly (so in practice doing `a + b` instead of `a.__add__(b)`), you would not need the checking for NotImplemented below?

jorisvandenbossche replied:
similar as https://github.com/pandas-dev/pandas/blob/c85ab083919b59ce84c220d5baf7d34ff4a0bcf2/pandas/core/indexes/category.py#L793

Dr-Irv replied:
The check for `NotImplemented` is needed because the underlying `ExtensionDtype` class (which are the objects in `lvalues` and `rvalues` above) may not implement the specific operator.

Not sure what you are referring to when you wrote "similar as" the code in `pandas/core/indexes/category.py".  That implementation is not doing things element by element, which is what I'm doing here.

jreback replied:
why are you doing it this way, dynamically generating this? there is a limited set of ops, just write them out.

jorisvandenbossche replied:
> why are you doing it this way, dynamically generating this? there is a limited set of ops, just write them out.

@jreback are you speaking about this `create_method` function that you commented on? Or the fact below the `create_method` function is used in a loop instead of manually repeated? 
Because if it is the first: that is how we do it almost everywhere in the pandas codebase.

jreback replied:
https://github.com/pandas-dev/pandas/pull/21191/commits/21d76b322c6e5ba87f95a9ac21a17e55b1915cc2

is what I mean.

Dr-Irv replied:
@jorisvandenbossche My latest commit does as you suggest.  This removes the tests for `NotImplemented`

@jreback My latest commit uses a similar pattern to what you had in the referenced commit, just creating an operator one by one instead of using a loop

jorisvandenbossche reviewed (commented):
some quick feedback

Dr-Irv committed:
fix lint issue and add ExtensionOpsMixin to API

Dr-Irv commented:
@jreback The idea is to use one piece of code for all of the operators, and to use the implementation of the operator on the underlying `ExtensionDtype`, **_if_** that implementation exists.

jbrockmendel commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +615,91 @@ def _ndarray_values(self):
+                    return f(b)
+                else:
+                    return NotImplemented
+            res = [callfunc(a, b) for (a, b) in zip(lvalues, rvalues)]
```
Orthogonal to @jorisvandenbossche 's comment, instead of `[callfunc(a, b) for (a, b) in zip(lvalues, rvalues)]` consider `libops.vec_binop(lvalues, rvalues, callfunc)`.  Should be equivalent but more performant.

jorisvandenbossche replied:
That doesn't seem to work properly?

```
In [64]: a = np.random.randn(100000).astype(object)

In [65]: b = np.random.randn(100000).astype(object)

In [66]: def func(a, b):
    ...:     return a + b
    ...: 
    ...: 

In [67]: a + b
Out[67]: 
array([-1.7373671804522872, -0.44608341233368065, -3.5963900176223706, ...,
       -1.0538754426986237, -2.0261407625013654, -0.6206810372160029], dtype=object)

In [68]: pd._libs.ops.vec_binop(a, b, func)
Out[68]: array([nan, nan, nan, ..., nan, nan, nan], dtype=object)

In [69]: np.array([func(aa, bb) for aa, bb in zip(a, b)], dtype=object)
Out[69]: 
array([-1.7373671804522872, -0.44608341233368065, -3.5963900176223706, ...,
       -1.0538754426986237, -2.0261407625013654, -0.6206810372160029], dtype=object)
```

jorisvandenbossche replied:
Ah, that's because it does a `maybe_convert_bool` at then end, I suppose. Is there a version that does not do this?

jbrockmendel replied:
Ahh, I guess that explains the "bin" part of "vec_binop".  So it looks like a) the original suggestion is incorrect and b) it would probably be easy to implement and may improve perf.

jorisvandenbossche replied:
I thought "binary" meant that it is an op between *2* values, not that the result in boolean. So I think the naming is just confusing.

jbrockmendel replied:
Agreed on both counts.

jbrockmendel commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +615,91 @@ def _ndarray_values(self):
         return np.array(self)
+
+
+def ExtensionOpsMixin(include_arith_ops, include_logic_ops):
```
Is there a compelling reason to make a factory for this instead of just making two mixin classes?

```
class ExtensionArithmeticMixin(...):

class ExtensionLogicMixin(...):
```
?

Dr-Irv replied:
@jbrockmendel I am sharing the code of the operator between arithmetic and comparison operators.  Alternatively, I could just create a base mixin class, and have `ExtensionArithmeticMixin` and `ExtensionLogicMixin` inherit from the base.

If you guys want me to make that change, let me know.

Dr-Irv replied:
@jbrockmendel I implemented your suggestion.  It does look cleaner.  It's in the latest commit

jbrockmendel commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +615,91 @@ def _ndarray_values(self):
+
+            # We can't use (NotImplemented in res) because the
+            # results might be objects that have overridden __eq__
+            if any(isinstance(r, type(NotImplemented)) for r in res):
```
I don't understand the comment above.  Is there a scenario in which `isinstance(r, type(NotImplemented)) and not (r is NotImplemented)`?

Dr-Irv replied:
My comment was about using `in`.  Having said that, I think I can use `(r is NotImplemented)` vs. `isinstance(r, type(NotImplemented))`

But now I have removed that in my working version based on @jorisvandenbossche suggestion to use the op rather than the op_name

jbrockmendel commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +615,91 @@ def _ndarray_values(self):
+        for op_name in arithops:
+            setattr(_ExtensionOpsMixin, op_name, create_method(op_name))
+
+    if include_logic_ops:
```
Can these be called `comparison_ops` instead of `logic_ops`?  I tend to think of the latter as meaning `&|^'

Dr-Irv replied:
Done

jbrockmendel commented on pandas/core/indexes/base.py:
```diff
@@ -2972,16 +2972,20 @@ def get_value(self, series, key):
             except KeyError:
-                if is_integer(key):
+                if (len(self) > 0 and
+                        self.inferred_type in ['integer', 'boolean']):
```
What makes this necessary?  I'm not clear on how it relates to the Mixin classes.

jorisvandenbossche replied:
For this you can comment on the relevant other PR (https://github.com/pandas-dev/pandas/pull/21260)

Dr-Irv replied:
@jbrockmendel the other PR (#21260) was needed to make the tests here work right.

jbrockmendel commented on pandas/core/ops.py:
```diff
@@ -990,6 +991,20 @@ def _construct_divmod_result(left, result, index, name, dtype):
     )
 
 
+def dispatch_to_extension_op(left, right, op_name):
```
Could this be accomplished with:

```
return dispatch_to_index_op(op, left, right, left.values.__class__)
```
?

jorisvandenbossche replied:
Unless we require that the ExtensionArray constructor handles objects of itself (and does not copy them), we cannot do the your example code (and currently we don't require this from extension array implementations I think. 
It is also don't think it necessarily improves readability of the code by reusing that.

jbrockmendel replied:
Fair enough.  Eventually dispatch_to_index_op is intended to give way to an EA-based dispatch.

Dr-Irv committed:
No longer use factory.  Use op instead of op_name. Catch exceptions

Dr-Irv commented:
@jbrockmendel @jreback @jorisvandenbossche My latest commit also does something with the JSON example to see that we trap different kinds of exceptions correctly.  I'm not sure if I should put that there or not.  It also illustrates that you can define just single methods using the mixin, i.e., if your underlying dtype only supports addition, but nothing else, you can then define an operator for addition and not define the other ones.

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +618,121 @@ def _ndarray_values(self):
+
+            return res_values
+
+        name = '__{name}__'.format(name=op_name)
```
why are you defining create_method as anything other than NotImplementedError

this is way too opionated - if u want to define a function to do this and an author can use this ok
but it should not be the default

maybe if u want to create an ExtensionOpsMixinForScalars (long name though)

jorisvandenbossche replied:
Please take a closer look at the diff if you give such an opinionated (pun intended :-)) comment

- this is not the default, it's an optional mixin extension authors can reuse (so what you suggest?)
- so it is exactly the point to be opinionated and fallback to the scalars (we might try to think of a better name that reflects this, as you suggested)
- it is what we discussed in the other PR

jreback replied:
@jorisvandenbossche maybe I wasn't clear.

``ExtensionOpsBase`` should simply have NotImplementedError for the actual implementation of the arithmethic and comaprison ops. Scalar implementations of this will IMHO rarely be used and simply cluttering things up. This can be put as a separate MixIn class if you wish for inclusion by an author, including it here is forcing a pretty bad idiom: coupling the interface and implementation, and is quite error prone because its not opt in. If an author extends and adds ops, they immediate get an implementation, which could very easiy be buggy / not work / work in unexpected ways.

I think the ops implementation is so important that this must be an explicit choice. 

The point of the mixin is to avoid the boilerplate of writing ``__div__ - ....`` and so on. NOT the implementation.

jorisvandenbossche replied:
> This can be put as a separate MixIn class if you wish for inclusion by an author, 

Again, this is what this PR is doing. 
To repeat from above, is might be it is just the name of ExtensionOpsBase which is not clear? (with which I agree)

Dr-Irv replied:
@jreback @jorisvandenbossche Maybe the naming is misleading.  So let me clarify the intent of the PR.

The goal here is that if someone extends `ExtensionArray` (let's call it `MyArray` for sake of discussion) where they are storing objects (let's call them `MyClass`) that already have the arithmetic and/or comparison operators defined for the class `MyClass`, they can use these mixins to get those operators defined on the `MyArray` subclass by simply adding the mixin as one of the base classes of `MyArray`.

An example is `Decimal`.  The operators are defined there, so the mixin gives you the operators on `DecimalArray` "for free".  I also have my own application that extends `ExtensionArray` and it uses the same principal.

The use of these 3 classes (`ExtensionOpsBase`, `ExtensionArithmeticMixin` and `ExtensionComparisonMixin`) is *optional* for people who are writing their own subclass of `ExtensionArray`.  The `ExtensionOpsBase` class is just a base class for the 2 mixin classes.  It contains the `create_method()` method that allows one to create an individual operator, under the assumption that `MyClass` has the operator defined.  Aside from the use of the two mixin classes that define all of the operators, `ExtensionOpsBase.createmethod()` can be used to define *individual* methods.  So, for example, if the `MyClass` class only supports addition, you can choose to not use the mixin, and just do something like the following in `MyArray` to create the operator.
```python
    __add__ = ExtensionOpsBase.create_method(operator.add)
```

IMHO, it seems that @jorisvandenbossche understands what I'm trying to do, but @jreback may not fully understand it.

If you think there is a better name for `ExtensionOpsBase`, I'm open to suggestions.

jorisvandenbossche replied:
To add, @jreback, if you don't like the exact implemenation with the three classes as it is now (and not the *idea* that @Dr-Irv explained above), we can certainly discuss that, no problem at all. 

I am also thinking if it couldn't be simplified. The ExtensionOpsBase is only there (or at least as a public class) in case you want to define *some* of them and not all, and thus cannot inherit the full ExtensionArithmeticMixin class. 
This is maybe an even more specific use case, so we can maybe leave that out for now. Although implementation wise, it's probably still the cleanest to have this class as a base class for the two mixin classes.

So for naming, we can maybe go for

```
ExtensionScalarOpsMixin
ExtensionScalarArithmeticMixin
ExtensionScalarComparisonMixin
```

? 

(so adding "scalar" in there, to indicate it uses the scalar values' ops, and rename the first one from Base to Mixin)

Dr-Irv replied:
@jorisvandenbossche I've tried a variety of implementations without the base class, and couldn't get it to work.  I'll rename as you suggest.

jreback replied:
renaming looks ok to me

jbrockmendel replied:
Agreed on the naming conventions.  Maybe the NotImplemented versions of the methods belong on the base `ExtensionArray`?

Dr-Irv replied:
@jbrockmendel The NotImplemented versions are there by default for any class, so I don't think we need to do anything.  In other words, if you write a class `MyClass` that extends `ExtensionArray` and you don't define the methods for operators, they are by default NotImplemented.

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +618,121 @@ def _ndarray_values(self):
+    operator using the class method create_method()
+    """
+    @classmethod
+    def create_method(cls, op):
```
can you make this a private method? (`_create_method`)

Dr-Irv replied:
@jorisvandenbossche See my reply here https://github.com/pandas-dev/pandas/pull/21261#discussion_r192418506 as to why I think it should not be private.

jorisvandenbossche replied:
It's fine to have it private for *users* (and I think that is better, as a end-user should never have to use this), but have it documented for extension array authors. We have other methods like that on the ExtensionArray class as well, like `_from_sequence`. 
We just need to clearly document it

Dr-Irv replied:
@jorisvandenbossche OK, that makes sense.  I'll make that change.

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +618,121 @@ def _ndarray_values(self):
+    Can also be used to define an individual method for a specific
+    operator using the class method create_method()
+    """
+    @classmethod
```
blank line above this for PEP8?

Dr-Irv replied:
It passed the PEP8 tests.......  But I can add it in.

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +618,121 @@ def _ndarray_values(self):
+    def create_method(cls, op):
+        """
+        A class method that returns a method that will correspond to an
+        operator for an ExtensionArray subclass.
```
mention that is to dispatch to the relevant operator defined on the scalars

Dr-Irv replied:
Done

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +618,121 @@ def _ndarray_values(self):
+        -----
+        Given an ExtensionArray subclass called MyClass, use
+
+        mymethod = create_method(my_operator)
```
I would give here an actual example (eg `__add__ = ExtensionOpsBase._create_method(operators.add)`)

Dr-Irv replied:
Done

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +618,121 @@ def _ndarray_values(self):
+        def _binop(self, other):
+            def convert_values(parm):
+                if isinstance(parm, ExtensionArray):
+                    ovalues = list(parm)
```
Why is it needed to convert to a list? ExtensionArrays support iterating through them (which returns the scalars)

Dr-Irv replied:
Fixed

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +618,121 @@ def _ndarray_values(self):
+            except TypeError:
+                msg = ("ExtensionDtype invalid operation " +
+                       "{opn} between {one} and {two}")
+                raise TypeError(msg.format(opn=op_name,
```
Personally I would not care catching the error. 
Eg for the decimal test case:

```
In [8]: import decimal

In [9]: d = decimal.Decimal('1.1')

In [10]: d + 'a'
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-10-c0db2fc01a1f> in <module>()
----> 1 d + 'a'

TypeError: unsupported operand type(s) for +: 'decimal.Decimal' and 'str'
```

which is already a nice error (and IMO even clearer, because it does not contain references to "Extension" which normal users don't necessarily are familiar with)

Dr-Irv replied:
Good point.  Fixed.

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +618,121 @@ def _ndarray_values(self):
+                                           one=type(lvalues),
+                                           two=type(rvalues)))
+
+            res_values = com._values_from_object(res)
```
I don't think this does anything for a list (it just passes it through), so this can be removed

Dr-Irv replied:
OK, I removed it and things still work.  Had copied this from elsewhere.

jorisvandenbossche commented on pandas/core/ops.py:
```diff
@@ -990,6 +991,26 @@ def _construct_divmod_result(left, result, index, name, dtype):
+
+    method = getattr(left.values, op_name, None)
+    if method is not None:
+        res_values = method(right)
```
Same comment as before, can we do here `op(left, right)` instead of getting the method manually ? (similar as what happens in `dispatch_to_index_op`) ?

Dr-Irv replied:
I've made that change.  Thanks.

jorisvandenbossche commented on pandas/core/ops.py:
```diff
@@ -990,6 +991,26 @@ def _construct_divmod_result(left, result, index, name, dtype):
+        res_values = method(right)
+    if method is None or res_values is NotImplemented:
+        msg = "ExtensionArray invalid operation {opn} between {one} and {two}"
+        raise TypeError(msg.format(opn=op_name,
```
given my comment above, this should not be needed (as the ExtensionArray op will raise the appropriate error)

Dr-Irv replied:
Yes, that makes sense.  I changed it.

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +618,121 @@ def _ndarray_values(self):
+
+            try:
+                res_values = self._from_sequence(res_values)
+            except TypeError:
```
would it makes sense to have a separate `_make_comparision_op` and `_make_aithmetics_op` ? 
The the comparison one should not need to try the conversion to ExtensionArray (if you want this to happen like in your case, you can still use the arithmetic one to add the comparison ops manually on your class)

Dr-Irv replied:
I see your point.  But if someone else had the same use case as me, where the comparison ops need to return an object rather than a boolean, then they'd have to know the workaround you suggest, and so then the question is whether we document it.

I'm leaving this as is for now, but will accept if you really want me to change it.

jorisvandenbossche replied:
One reason to split it up would be to avoid the extra overhead to try to convert it to an ExtensionArray in case of the boolean ops (although for a good implementation this probably should not be too much overhead ..)

Dr-Irv replied:
OK, I've done the split for the next commit.  I added a parameter to `create_method()` to specify whether to try the coercion to the underlying dtype, and said to not do it for the comparison operators.

jorisvandenbossche reviewed (commented):
Thanks for the updates! 

(didn't look at the tests yet)

jreback commented:
This needs some tests, seee: https://github.com/pandas-dev/pandas/pull/21160/files#diff-f12c8797eb9d98e36fd196d5e0821f75  for a generic way to do this.

FInally I would suggest that the implementation of Categorical use this (can be a followup PR) to validate this approach. 

Too much code has been added w/o adequate testing.

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -14,6 +14,7 @@ Other Enhancements
 ^^^^^^^^^^^^^^^^^^
 - :func:`to_datetime` now supports the ``%Z`` and ``%z`` directive when passed into ``format`` (:issue:`13486`)
 - :func:`to_csv` now supports ``compression`` keyword when a file handle is passed. (:issue:`21227`)
+- ``ExtensionArray`` has a ``ExtensionOpsMixin`` factory that allows default operators to be defined (:issue:`20659`, :issue:`19577`)
```
make a separate sub-section. pls expand this a bit, I know what you mean, but I doubt the average reader does.

jbrockmendel replied:
Is the note here still accurate or has the factory been changed to just mixins?

Dr-Irv replied:
I've written a section for whatsnew

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +618,121 @@ def _ndarray_values(self):
         return np.array(self)
+
+
+class ExtensionOpsBase(object):
```
this should not contain a scalar based implemetation, if you really want to add that make it a separate MixIin (inherit from object)

jorisvandenbossche replied:
Note this is currently *not* a "base" class as what you have in mind, so it might be you just find the naming wrong?

jreback replied:
it IS a Base class for the arithemetic / comparison ops. This is the wrong place to put this logic. This needs to return NotImplementeError. as I have said multiple times.

Dr-Irv replied:
@jreback see my comment https://github.com/pandas-dev/pandas/pull/21261#discussion_r192418506 .  The class `ExtensionOpsBase` (and I'm open to new names) is not meant to be used directly as a parent class.  So why would it need to return NotImplementedError?  In fact, if someone just puts `ExtensionOpsBase` as a parent, it will do nothing beyond defining `create_method()`.

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +618,121 @@ def _ndarray_values(self):
+        return set_function_name(_binop, name, cls)
+
+
+class ExtensionArithmeticMixin(ExtensionOpsBase):
```
inherit from object

jorisvandenbossche replied:
You will need to be clearer what your objection is. In the way @Dr-Irv implemented it, he needs to subclass the other class, to have the `create_method` implementation. If you don't like the way how it is currently split in three classes, be specific about that.

jreback replied:
same point as above ``create_method`` needs to return NotImplementedError in this class, hence you don't need a base class. I am being pedantic on purpose here. What you are doing is coupling way to tight the scalar implementation.

Dr-Irv replied:
@jreback can't inherit from object, as the `set_function_name` call in `create_method()` needs to know the parent class.  (or maybe you know some other way to do this)

I don't see the point that `create_method` needs to return NotImplementedError, as the whole point of what I'm doing here is to just provide the capability to have the operators on `ExtensionArray` use the underlying operators of the elements on the array.

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +618,121 @@ def _ndarray_values(self):
+        __div__ = ExtensionOpsBase.create_method(operator.div)
+        __rdiv__ = ExtensionOpsBase.create_method(ops.rdiv)
+
+    __divmod__ = ExtensionOpsBase.create_method(divmod)
```
rdivmod is a thing

Dr-Irv replied:
Fixed.

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -2972,16 +2972,20 @@ def get_value(self, series, key):
-                if is_integer(key):
+                if (len(self) > 0 and
+                        self.inferred_type in ['integer', 'boolean']):
+                    raise
```
pls move all non-ops code to another PR

jorisvandenbossche replied:
The same answer as above: this is needed for the test to pass untill the other PRs are merged (to say: it is already moved to separate PRs)

jreback commented on pandas/core/series.py:
```diff
@@ -2221,26 +2220,36 @@ def combine(self, other, func, fill_value=np.nan):
         Series.combine_first : Combine Series values, choosing the calling
             Series's values first
         """
+        self_is_ext = is_extension_array_dtype(self.values)
```
I would move all of this (combine stuff) to another PR on top of this one.

jorisvandenbossche replied:
https://github.com/pandas-dev/pandas/pull/21183

jreback commented on pandas/core/series.py:
```diff
@@ -2221,26 +2220,36 @@ def combine(self, other, func, fill_value=np.nan):
+                new_values = [func(lv, other) for lv in self._values]
             new_name = self.name
+
+        if self_is_ext and not is_categorical_dtype(self.values):
```
don't use shorthands like this

make this simpler, something like
```
if is_categorical_dtype(...):
    pass
elif is_extension_dtype(...):
   ....
```

much more readable

jorisvandenbossche replied:
Comment that in https://github.com/pandas-dev/pandas/pull/21183

jreback commented on pandas/core/series.py:
```diff
@@ -2221,26 +2220,36 @@ def combine(self, other, func, fill_value=np.nan):
+        if self_is_ext and not is_categorical_dtype(self.values):
+            try:
+                new_values = self._values._from_sequence(new_values)
+            except TypeError:
```
why are you catching a TypeError?

jorisvandenbossche replied:
Is also for the other PR, but: because the result might not necessarily be an ExtensionArray.

This is of course a bit an unclear area of `combine`, but currently there is no restriction on the function that the dtype needs to be preserved in the result of the function.

jreback commented on pandas/tests/extension/decimal/array.py:
```diff
@@ -24,11 +26,14 @@ def construct_from_string(cls, string):
-        assert all(isinstance(v, decimal.Decimal) for v in values)
+        for val in values:
+            if not isinstance(val, self.dtype.type):
+                raise TypeError
```
erorr message

Dr-Irv replied:
fixed (and also in pandas/tests/extension/json/array.py)

jreback commented on pandas/tests/extension/decimal/test_decimal.py:
```diff
@@ -183,3 +186,36 @@ def test_dataframe_constructor_with_different_dtype_raises():
+
+
+class TestOperator(BaseDecimal, TestSeriesOperators):
+    @cache_readonly
```
never do this. use a fixture

Dr-Irv replied:
This is unavoidable, because I am subclassing the `tests.series.tests_operators.TestSeriesOperators` class.  That class inherits from `tests.series.common.TestData`, which defines `ts` as a `@cache_readonly` property.  The method I use for testing (`TestSeriesOperators.test_operators()`) has references to `self.ts`, i.e., it doesn't use a fixture, and I can't see how to override that definition so we get a `DecimalArray` series and reuse all that testing code.

Dr-Irv replied:
Never mind, I redid the tests as you suggested elsewhere

jreback commented on pandas/tests/extension/decimal/test_decimal.py:
```diff
@@ -183,3 +186,36 @@ def test_dataframe_constructor_with_different_dtype_raises():
+                return pd.Series(vals._from_sequence([abs(i) for i in vals]))
+            else:
+                return abs(v)
+        context = decimal.getcontext()
```
what the heck is all this?

Dr-Irv replied:
I am reusing the tests that are in the class  `pandas/tests/series/test_operators/TestSeriesOperators`.  For decimal, the default when dividing by zero is to raise an error, as opposed to in numpy, where `inf` is returned.  So by changing the defaults for `Decimal`, we get the same behavior as numpy.

There is a related issue with respect to `abs()`, because the tests in `TestSeriesOperators` call `np.abs()`, but that doesn't work for the `Decimal` class.

jreback commented on pandas/tests/extension/json/test_json.py:
```diff
@@ -230,3 +230,21 @@ def test_groupby_extension_agg(self, as_index, data_for_grouping):
         )
+
+
+def test_ops(data):
```
rather than try to create one off tests, you need a genearlized testing mechanism dispatched in BaseOps

Dr-Irv replied:
OK, I copied your code and made some modifications to it to have a general testing mechanism.

jreback reviewed (changes requested):
needs quite a bit of splitting / work

Dr-Irv committed:
Numerous changes based on PR feedback

pep8speaks commented:
Hello @Dr-Irv! Thanks for updating the PR.















 - In the file [`pandas/tests/extension/base/__init__.py`](https://github.com/pandas-dev/pandas/blob/ec96841d0c4b9a2bb8a2c52c96de10d3696a429c/pandas/tests/extension/base/__init__.py), following are the PEP8 issues :

> [Line 50:30](https://github.com/pandas-dev/pandas/blob/ec96841d0c4b9a2bb8a2c52c96de10d3696a429c/pandas/tests/extension/base/__init__.py#L50): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment

Dr-Irv commented:
@jreback @jorisvandenbossche @jbrockmendel I have just pushed a new commit that addresses a number of issues raised above, including:
  - Updated whatsnew to provide more of a description
  - Renamed the mixin classes
  - Let errors generated by missing implementations of the underlying classes get reported as is
  - Implement tests based on test framework written by @jreback elsewhere.

Dr-Irv committed:
PEP8 and merge conflicts

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +617,124 @@ def _ndarray_values(self):
+            def convert_values(parm):
+                if isinstance(parm, ExtensionArray) or is_list_like(parm):
+                    ovalues = parm
+                elif is_extension_array_dtype(parm):
```
A Series is also 'list-like', so I think this `elif` will not be reached. 
But given that you simply iterate through it (which will be the same for series or extensionarray), I think you can simply remove that part.

Dr-Irv replied:
Done

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +617,124 @@ def _ndarray_values(self):
+                else:  # Assume its an object
+                    ovalues = [parm] * len(self)
+                return ovalues
+            lvalues = convert_values(self)
```
This should not be needed as the calling object will always be already an ExtensionArray?

Dr-Irv replied:
Yes.  Fixed

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +617,124 @@ def _ndarray_values(self):
+        op_name = ops._get_op_name(op, False)
+
+        def _binop(self, other):
+            def convert_values(parm):
```
minor detail, but I find 'values' (or 'param') is clearer than 'parm'

Dr-Irv replied:
Changed

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +617,124 @@ def _ndarray_values(self):
+
+            return res
+
+        name = '__{name}__'.format(name=op_name)
```
you can also pass True to `_get_op_name`, and then that function does this formatting for you

Dr-Irv replied:
Changed

jorisvandenbossche commented on pandas/tests/extension/base/ops.py:
```diff
@@ -0,0 +1,66 @@
+
+
+class BaseOpsTests(BaseExtensionTests):
+    """Various Series and DataFrame ops methos."""
```
methos -> methods

Dr-Irv replied:
fixed

jorisvandenbossche commented on pandas/tests/extension/base/ops.py:
```diff
@@ -0,0 +1,66 @@
+    def check_op(self, s, op, other, exc=NotImplementedError):
+
+        with pytest.raises(exc):
+            getattr(s, op)(other)
```
IMO we should rather test `op(s, other)`, but this is because the fixture gives the dunder ops?

jorisvandenbossche replied:
Ah, I see you do that below for the decimal tests. 
But, I might maybe rather put that implementation here as the default, and overwrite it in categorical to check it raises instead?

jorisvandenbossche replied:
Hmm, although, then the json tests also needs to skip it. Maybe we can have both versions here in the base tests, and then you can subclass the appropriate one (I don't know how typical it would be external authors would be able to reuse the base class for those tests though)

Dr-Irv replied:
@jorisvandenbossche I copied this from implementation done by @jreback elsewhere.  I actually agree with you that we should use the op here rather than the dunder name.  I will change the base class and see how things look.

jorisvandenbossche commented on pandas/tests/extension/base/ops.py:
```diff
@@ -0,0 +1,66 @@
+        # scalar
+        op = all_arithmetic_operators
+        s = pd.Series(data)
+        self.check_op(s, op, 1, exc=TypeError)
```
maybe use `s[0]` as the scalar? (so to have the same type) That might be more general likely to work.

jorisvandenbossche commented on pandas/tests/extension/base/ops.py:
```diff
@@ -0,0 +1,66 @@
+    def _compare_other(self, data, op, other):
+        s = pd.Series(data)
+
+        if op in '__eq__':
```
in -> == ?

But it's not fully clear to me what you are testing here. You are testing that the equality operators are *not* implemented?

Dr-Irv replied:
This is from the @jreback implementation.  Let me study it in more detail and try it another way.

Dr-Irv replied:
@jorisvandenbossche So I studied this, and I think the intent here is to say that by default, the `ExtensionArray` `eq` and `ne` operators are not implemented, but that doing an `eq` or `ne` operation on a `Series` containing the `ExtensionArray` produces a result.

But I have made some changes to these tests to clarify that we are testing operator names versus operators.

jorisvandenbossche commented on pandas/tests/extension/base/ops.py:
```diff
@@ -0,0 +1,66 @@
+from .base import BaseExtensionTests
+
+
+class BaseOpsTests(BaseExtensionTests):
```
maybe we can split this in arithmetic and compare ops? So it is easier to fully skip one of the two, depending on what your class implements.

Dr-Irv replied:
I will do that.

jorisvandenbossche commented on pandas/tests/extension/decimal/test_decimal.py:
```diff
@@ -183,3 +184,64 @@ def test_dataframe_constructor_with_different_dtype_raises():
+        # scalar
+        op_name = all_arithmetic_operators
+        s = pd.Series(data)
+        self.check_op(s, op_name, decimal.Decimal(1.5))
```
if you update the scalar in the base test, I think this one can be inherited?

Dr-Irv replied:
Yes, that was possible

jorisvandenbossche commented on pandas/tests/series/test_operators.py:
```diff
@@ -1216,11 +1216,11 @@ def test_neg(self):
         assert_series_equal(-(self.series < 0), ~(self.series < 0))
 
-    def test_operators(self):
+    def test_operators(self, absfunc=np.abs):
```
is this change related to the PR?

Dr-Irv replied:
This was leftover from the previous method of testing, so I will revert it back.

jorisvandenbossche reviewed (commented):
Thanks for the updates! 
Added some more comments

TomAugspurger commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -10,6 +10,27 @@ New features
+``ExtensionArray`` operator support
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+An ``ExtensionArray`` subclass consisting of objects that have arithmetic and
```
Can you include most of this in `extending.rst` as well?

It'd be good to mention the performance implications of using these mixins.

Dr-Irv commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +617,124 @@ def _ndarray_values(self):
+
+        Parameters
+        ----------
+        op:               An operator that takes arguments op(a, b)
```
Done

Dr-Irv committed:
Merge remote-tracking branch 'upstream/master' into opsmixin

Dr-Irv committed:
Update docs and reorganize tests

Dr-Irv commented:
@jorisvandenbossche @TomAugspurger Latest commit reorganizes the tests (based on your suggestions), adds docs to extending.rst, and most importantly, fixes a bug in how the reverse operators were called.

jreback commented on doc/source/extending.rst:
```diff
@@ -113,6 +113,37 @@ by some other storage type, like Python lists.
 See the `extension array source`_ for the interface definition. The docstrings
 and comments contain guidance for properly implementing the interface.
 
+:class:`~pandas.api.extensions.ExtensionArray` Operator Support
```
needs a ref

Dr-Irv replied:
done

jreback commented on doc/source/extending.rst:
```diff
@@ -113,6 +113,37 @@ by some other storage type, like Python lists.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, there are no operators defined for the class :class:`~pandas.api.extensions.ExtensionArray`.
+There are two ways that you can provide operator support for your ExtensionArray. One
```
can you use bullets / numbers  here instead (on the 2 items)

Dr-Irv replied:
done

jreback commented on doc/source/extending.rst:
```diff
@@ -113,6 +113,37 @@ by some other storage type, like Python lists.
+element one-by-one, this might not be as performant as implementing your own
+version of the associated operators directly on the ExtensionArray.
+
+Testing Extension Arrays
```
can you add a ref to this section too

Dr-Irv replied:
done

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -10,6 +10,27 @@ New features
 
 .. _whatsnew_0240.enhancements.other:
 
```
add a ref here

Dr-Irv replied:
done

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -10,6 +10,27 @@ New features
+An ``ExtensionArray`` subclass consisting of objects that have arithmetic and
+comparison operators defined on the underlying objects can easily support 
+those operators on the ``ExtensionArray``, and therefore the operators 
+on ``Series`` built on those ``ExtensionArray`` classes will work as expected.
```
same comment as above (use bullets / numbers)

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -10,6 +10,27 @@ New features
+on ``Series`` built on those ``ExtensionArray`` classes will work as expected.
+
+Two new mixin classes, :class:`ExtensionScalarArithmeticMixin` and
+:class:`ExtensionScalarComparisonMixin`, support this capability.  
```
put a ref to the extending.rst doc section for this (e.g. see docs)

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -7,8 +7,14 @@
+from pandas.compat import set_function_name, PY3
+from pandas.core.dtypes.common import is_list_like
+from pandas.core import ops
+from pandas.core.ops import _get_op_name
```
can this last be just used as ``ops._get_op_name`` (when you use it)

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +616,122 @@ def _ndarray_values(self):
+        return set_function_name(_binop, op_name, cls)
+
+
+class ExtensionScalarArithmeticMixin(ExtensionScalarOpsMixin):
```
still not clear to why this inherits ``ExtensionScalarOpsMixin``. If I want *only* the ops definitions, and want to define my own create_method this seems a bit odd to do. Rather I would like to have a ``_create_method`` that just raises NotImplementedError, so things fail as soon as I inherit this base class. 

the way this is written it is entirely too error prone. You drop in these classes, then it *might* work incorrectly! e.g. maybe I happen to have a scalar ops defined. This should be very explict. I dont' see ANY advantage of defaulting to a scalar ops schema here.

Dr-Irv replied:
@jreback I think I see what you are saying, so I have pushed a commit that I think does what you suggest.  I've created infrastructure that allows a developer to define their own `create_method`. I've written up docs that explains how it works.  The underlying implementation is a little complex, because there are lots of inner classes flying around, but now people can use the `ExtensionScalarArithmeticMixin` if they have the underlying objects with operators defined, or they can use `ExtensionArithmeticMixin(MyMixin)` to define all of the operators, and then the class `MyMixin` has the definition of `create_method` in it.

I hope this addresses your concern.

jorisvandenbossche replied:
Personally, I find this much more complicated than what you had before, and I am also not really sure if it will be that useful in practice (i.e. the usecase of wanting to implement your own `_create_method` but that works for all operators. Which actually was also possible with your previous approach by inheriting eg ExtensionScalarArithmeticMixin and overriding the `_create_method` method). 
I would rather wait and see if there is need for supporting this usecase.

jorisvandenbossche replied:
> I dont' see ANY advantage of defaulting to a scalar ops schema here.

@jreback again, 1) it's *not* defaulting for the generic ExtensionArray, the scalar ops is only implemented for this *optional* mixin 2) this is entirely the point of this PR. 
Yes, you may well never want to use such fallback to the scalars. But then you just don't use those mixins from this PR.

This PR is not intending to solve the general ops issue for ExtensionArrays, its purpose was to provide a easy solution for a very specific use case.

jreback reviewed (changes requested):
I haven't fully reviewed but some comments.

jorisvandenbossche commented:
Can you take a look at https://github.com/pandas-dev/pandas/pull/21183 (then the diff here will be smaller), both me and Tom approved it, but since you actually commented on that code but in this PR, please take a look there.

jorisvandenbossche commented:
@jreback Can you take a look at https://github.com/pandas-dev/pandas/pull/21183 (then the diff here will be smaller), both me and Tom approved it, but since you actually commented on that code but in this PR, please take a look there.

Dr-Irv committed:
Allow users to define their own operators. Additional docs

jorisvandenbossche commented on doc/source/extending.rst:
```diff
@@ -113,21 +113,67 @@ by some other storage type, like Python lists.
+There are two approaches for providing operator support for your ExtensionArray:
+
+1. Define each of the operators on your ExtensionArray subclass. 
+2. Use operators from pandas defined on the ExtensionArray subclass based on already defined
```
I would remove "use operators from pandas ..". The essence is "operators defined on the underlying elements (scalars) of the ExtensionArray"

Dr-Irv replied:
How about: "Use an operator implementation from pandas that depends on operators that are already defined on the underlying elements (scalars) of the ExtensionArray."

jorisvandenbossche commented on doc/source/extending.rst:
```diff
@@ -113,21 +113,67 @@ by some other storage type, like Python lists.
+2. Use operators from pandas defined on the ExtensionArray subclass based on already defined
+   operators on the underlying elements.
+    
+For the first approach, you will need to create a mixin class with a single class method,
```
I would not explain it like this. A very likely approach for the first case is that you actually define the operators you want to support by adding `__add__`, `__eq__` implementations etc.

Dr-Irv replied:
I think this is the heart of what @jreback was getting at, i.e., would one create each operator independently, or create one `_create_method()` implementation that takes care of all of the operators.

jorisvandenbossche commented on doc/source/extending.rst:
```diff
@@ -113,21 +113,67 @@ by some other storage type, like Python lists.
+
+    class MyOpsMixin(object):
+    @classmethod
+    def _create_method(cls, op, coerce_to_dtype=True):
```
extra level of indentation for valid python code

Dr-Irv replied:
fixed

jorisvandenbossche commented on doc/source/extending.rst:
```diff
@@ -113,21 +113,67 @@ by some other storage type, like Python lists.
+The second approach assumes that the underlying elements of the ExtensionArray 
+have the individual operators already defined.  In other words, if your ExtensionArray
+named ``MyExtensionArray`` is implemented so that each element is an instance 
+of the class ``MyExtensionElement``, then if the operators are defined 
```
I would somewhere use "scalar type" here, as that is the terminology we use above

Dr-Irv replied:
added

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -679,13 +710,105 @@ def convert_values(param):
 
 
-class ExtensionScalarArithmeticMixin(ExtensionScalarOpsMixin):
+def ExtensionArithmeticOpsMixin(base=ExtensionDefaultOpsMixin):
```
Why is it needed to make this a factory function? 
The alternative is make this a class and then the user needs to implement a `_create_method` in its ExtensionArray ?

Dr-Irv replied:
@jorisvandenbossche I don't see an easy way of sharing the implementation of `_create_method()` that is in `ExtensionScalarOpsMixin` with the two mixin classes `ExtensionScalarArithmeticMixin` and `ExtensionScalarComparisonMixin`.  I think it's because `_create_method()` is a class method and when `ExtensionArithmeticOpsMixin` is set up, it needs to know the base class that contains the class method `_create_method()`.

The alternative is to do it how I did it before - i.e., just make `ExtensionScalarOpsMixin` a base class.

jorisvandenbossche reviewed (commented):
As I said in my previous comment, I am not really sure about the added value of this more complex implementation, or at least for now.

My take: let's focus in this PR on what you actually wanted: a mixin that you can inherit if you want default scalar-ops-fallback behaviour, and leave more generic ways to add ops for another PR, as that needs more discussion.

Dr-Irv commented:
@jorisvandenbossche You wrote:

> As I said in my previous comment, I am not really sure about the added value of this more complex implementation, or at least for now.
> 
> My take: let's focus in this PR on what you actually wanted: a mixin that you can inherit if you want default scalar-ops-fallback behaviour, and leave more generic ways to add ops for another PR, as that needs more discussion.

I'm in agreement with you.  I was responding to what I think was on the mind of @jreback.  I'm not sure how to reconcile your position with his.

Having said that, the hooks (`dispatch_to_extension_op()` that I have put in pandas/core/ops.py will work independent of how a developer implements the ops on the ExtensionArray subclass. 

IMHO, I view the original implementation in the PR as a stepping stone towards having more generic ops support for EA's.

jreback commented:
well I have state my point many times. I would simply drop the scalar implementation entirely. I agree its getting too complex. If an extension class wants to implement ops, then let them. But providing scalar support is a pretty specific usecase IMHO (and to be honest not very general).

jreback commented:
@Dr-Irv your implemenation is way too complex here.

Just make 2 classes. Define the actual operators

```
class _ExtensionArithmeticOpsMixin(base):
+        __add__ = base._create_method(operator.add)
+        __radd__ = base._create_method(ops.radd)
+        __sub__ = base._create_method(operator.sub)
+        __rsub__ = base._create_method(ops.rsub)
+        __mul__ = base._create_method(operator.mul)
+        __rmul__ = base._create_method(ops.rmul)
+        __pow__ = base._create_method(operator.pow)
+        __rpow__ = base._create_method(ops.rpow)
+        __mod__ = base._create_method(operator.mod)
+        __rmod__ = base._create_method(ops.rmod)
+        __floordiv__ = base._create_method(operator.floordiv)
+        __rfloordiv__ = base._create_method(ops.rfloordiv)
+        __truediv__ = base._create_method(operator.truediv)
+        __rtruediv__ = base._create_method(ops.rtruediv)
+        if not PY3:
+            __div__ = base._create_method(operator.div)
+            __rdiv__ = base._create_method(ops.rdiv)
+
+        __divmod__ = base._create_method(divmod)
+        __rdivmod__ = base._create_method(ops.rdivmod)
```
and that's it.

If you want to extend and do scalar type artihmetic in our own extension class then by all means.
Not sure any reason for anything beyond basic ops definitions should live here.

Dr-Irv commented:
@jreback You wrote:

> @Dr-Irv your implemenation is way too complex here.
> 
> Just make 2 classes. Define the actual operators
> 

> ```
> class _ExtensionArithmeticOpsMixin(base):
> +        __add__ = base._create_method(operator.add)
> +        __radd__ = base._create_method(ops.radd)
> +        __sub__ = base._create_method(operator.sub)
> ## Omitting rest
> ```

The problem here is that `_create_method()` needs to be a class method of the parent class.  So you need to know the parent class.  I don't see how your idea will work.

I'm going to revert back to the prior implementation, where there is a base `ExtensionScalarOpsMixin` and `ExtensionScalarArithmeticMixin` and `ExtensionScalarComparisonMixin` inherit from that.

jorisvandenbossche commented:
> well I have state my point many times. I would simply drop the scalar implementation entirely. 

The scalar implementation is what this PR was proposing from the beginning, and AFAIK you agreed on providing, for those users that want it, an *optional* mixin that provides scalar-ops fallback behaviour. Yes, you made clear comments about not liking the *implementation*, but not about the starting point of this PR. 
This PR was not about adding a generic ops implementation (as you have a PR for that in https://github.com/pandas-dev/pandas/pull/21191), but purely the scalar ops mixin. Let's just keep it separated that way, and discuss the generic ops in a separate issue/PR.

jorisvandenbossche commented:
> The problem here is that _create_method() needs to be a class method of the parent class. So you need to know the parent class. I don't see how your idea will work.

Can't this be `self`?

jbrockmendel commented:
@jreback more or less in agreement with @jorisvandenbossche here, though my main thought is that the scalar-dispatch is effectively what will be needed for object-dtyped arrays.

Dr-Irv commented:
I wrote:
> The problem here is that _create_method() needs to be a class method of the parent class. So you need to know the parent class. I don't see how your idea will work.

@jorisvandenbossche 
>  Can't this be self?

No, because we want `__add__`, `__sub__`, etc. to be defined on the subclass of the `ExtensionArray` _class_, not on each instance of the `ExtensionArray` subclass.

For example, using `DecimalArray`, we want `__add__` to be defined on `DecimalArray`, and since `_create_method()` contains the code to create the  method, `_create_method()` is a class method.

jreback commented:
as i did in my impl

this is so much simpler to just make class methods

then you opt in by 

MyExtensionArray._add_numeric_ops()

and just override the _create_method
(or even could just pass it here directly)

Dr-Irv commented:
@jreback YOu wrote:
> as i did in my impl
> 
> this is so much simpler to just make class methods
> 
> then you opt in by
> 
> MyExtensionArray._add_numeric_ops()
> 
> and just override the _create_method
> (or even could just pass it here directly)

OK, but that's not a true mixin.  I'll do as you suggest.

jorisvandenbossche commented:
@jreback so what you propose (from a user perspective) is to have a user do

```
class MyExtensionArray(ExtensionArray, ExtensionScalarOpsMixin)
    pass

MyExtensionArray._add_arithmetic_ops()
MyExtensionArray._add_comparison_ops()
```

instead of the current

```
class MyExtensionArray(ExtensionArray, 
                       ExtensionScalarArithmeticOpsMixin,
                       ExtensionScalarComparisonOpsMixin)
    pass

```

to be clear on the API you propose?

Dr-Irv committed:
Use approach of initializing operators via func calls

Dr-Irv commented:
@jbrockmendel wrote:

> @jreback more or less in agreement with @jorisvandenbossche here, though my main thought is that the scalar-dispatch is effectively what will be needed for object-dtyped arrays.

Yes, that is exactly my use case!

For @jreback and @jorisvandenbossche , latest commit does it the way that @jreback suggests

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +615,118 @@ def _ndarray_values(self):
+    A base class for linking the operators to their dunder names
+    """
+    @classmethod
+    def addArithmeticOps(cls):
```
please use PEP8 and make it private (`_add_arithmetic_ops`)

Dr-Irv replied:
Done.  Somehow flake8 didn't pick it up for me when I ran it this time.

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -617,42 +617,60 @@ def _ndarray_values(self):
-        def returnNotImplemented(self, other):
-            return NotImplemented
+    def addArithmeticOps(cls):
+        cls.__add__ = cls._create_method(operator.add)
```
name these like

_add_numeric_ops and so on

Dr-Irv replied:
Fixed

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +615,118 @@ def _ndarray_values(self):
+        cls.__lt__ = cls._create_method(operator.lt, False)
+        cls.__gt__ = cls._create_method(operator.gt, False)
+        cls.__le__ = cls._create_method(operator.le, False)
+        cls.__ge__ = cls._create_method(operator.ge, False)
```
The `False` here is quite specific to the `_create_method` implementation of the ScalarOpsMixin, so this doesn't feel completely "clean". 

So you could also have a `_create_arithmetic_method` and `_create_comparison_method` class methods, and in your scalar mixing below you can reuse the same function for both those class methods (with a different value of the keyword).

jorisvandenbossche reviewed (commented):
For a user perspective, I personally slightly prefer the mixin classes to having to call the class method. But if this is a way we can agree on to move forward, then that's fine!

Dr-Irv committed:
Fix lint errors and make add methods private

jorisvandenbossche commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -10,6 +10,45 @@ New features
+For the second approach, which is appropriate if your ExtensionArray contains 
+elements that already have the operators
+defined on a per-element basis, pandas provides a mixin,
+:class:`ExtensionScalarOpsMixin` that you can use that can
```
"that you can use that can define" -> "that you can use to define"

Dr-Irv replied:
fixed

jorisvandenbossche commented on doc/source/extending.rst:
```diff
@@ -113,6 +113,52 @@ by some other storage type, like Python lists.
+for ``MyExtensionElement``, the second approach will automatically
+define the operators for ``MyExtensionArray``.
+
+A mixin class, :class:`~pandas.api.extensions.ExtensionScalarOpscMixin` supports this second
```
typo in "Opsc"

Dr-Irv replied:
fixed

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +615,125 @@ def _ndarray_values(self):
+    insert the lines
+
+    MyExtensionArray.addArithmeticOperators()
+    MyExtensionArray.addComparisonOperators()
```
left-over of the old name

jorisvandenbossche commented on pandas/core/ops.py:
```diff
@@ -990,6 +991,26 @@ def _construct_divmod_result(left, result, index, name, dtype):
+    else:
+        # We know that left is not ExtensionArray and is Series and right is
+        # ExtensionArray.  Want to force ExtensionArray op to get called
+        res_values = op(list(left.values), right.values)
```
was this needed to fix failing tests? (this was not here in a previous version I think? and eg the dispatch to index is only dealing with the left case)

Dr-Irv replied:
@jbrockmendel Wrote:
> >  Another possible implementation might be to add a test in the wrappers that says that if is_extension_dtype(right) return NotImplemented

>This is much closer to what we've had in mind with the recent refactoring in ops. Would this look something like splitting up the conditions on 1082-1085? We might need to up-cast right to be a Series in order for the deference chain to go through; I'll need to double-check.

Yes, I think it would be something like:
```
       elif is_extension_array_dtype(left):
           return dispatch_to_extension_op(op, left, right)
       elif (is_extension_array_dtype(right) and                 
              not is_categorical_dtype(right)):
           return NotImplemented
```
But I would have to test the concept.  I don't think the upcast is necessary. We know that left is a Series. So python would then see that `op(left, right)` returns `NotImplemented`, and then try to call `reverse_op(right, left)`, so that the reverse operator is called, but using the implementation of `right`.  

I'm not going to try making that change without feedback from others.

> Why is the list(left.values) necessary? I would have thought left.values would get the job done.

Because `left.values` could be any type (numpy.ndarray, pd.Categorical, etc.), and we don't want the version of op to be called that is associated with the class of left.values . So by making it a list, we are forcing the reverse operator to be called on right.values, which is the EA implementation.

jorisvandenbossche replied:
> Another possible implementation might be to add a test in the wrappers that says that if is_extension_dtype(right) return NotImplemented, which will then make python call the reverse operator.

I think this is the better way to go. 

For me it is fine to not yet do this in this PR, but then I would also not include the above change (and live with that the first iteration in this PR is not yet ideal for cases where the ExtensionArray is the right variable)

Dr-Irv replied:
@jorisvandenbossche If I don't handle the  case where ExtensionArray is the right variable now, then my tests will fail.  So I either implement it the other way (as you quoted above), or leave it as I have it now.  Should I make that change (using `NotImplemented`)?

jorisvandenbossche replied:
> Should I make that change (using NotImplemented)?

I would at least try this

Dr-Irv replied:
@jorisvandenbossche Yes, I will try this in a new PR.

Dr-Irv replied:
@jorisvandenbossche (and @jbrockmendel) I tried the `NotImplemented` idea and it doesn't work.  I think the reason is as follows.  Let's say that we have one `Series` called `s1` that has a normal (`int` or `double`) dtype. And `s2` has a dtype of the EA.  Let's say you write the expression `s1 + s2`.  Python sees that you are adding two `Series` objects.  So internally, if we return `NotImplemented` because `s2` is an EA, then Python will say "Well, you don't know how to add a Series to a Series, so I'm not going to bother calling the reverse operator."  In other words, the reverse operator is only called when the two types passed to the operator are different, and `NotImplemented` is returned when applying the operator to the first argument.

jbrockmendel replied:
I'm not entirely sure I follow.  Are you saying that both `s1.__add__(s2)` and `s2.__add__(s1)` return `NotImplemented`?

The dispatch logic should do something like
```
out = s1.values + s2
return Series(out, name=...)
```

It is `s1.values.__add__` that should return `NotImplemented` in this case.  Is this not an option for some reason?

Dr-Irv replied:
@jbrockmendel I've tried two different alternatives (regarding replacing this code
```python
else:
        # We know that left is not ExtensionArray and is Series and right is
        # ExtensionArray.  Want to force ExtensionArray op to get called
        res_values = op(list(left.values), right.values)
```
that this discussion refers to). Note the section of code is hit when `left.values` is *not* an EA, but `right.values` is an EA.
1. Use
```python
else:
    return NotImplemented
```
2. Use
```python
else:
    res_values = op(left.values, right.values)
```

In the first case, python sees that `s1.__add__(s2)` is `NotImplemented`, and because `s1` and `s2` are both of type `Series`, then `s2.__radd__(s1)` is never called.

In the second case, `left.values` is a `numpy` array, so then the `numpy` `__add__` method is called, and it is perfectly happy to return an array of `dtype` that is set to `object` rather than the dtype of the EA.  `numpy` ends up calling the `__add__` method for each of the elements.

jorisvandenbossche reviewed (commented):
some quick comments, didn't look in detail again

Dr-Irv committed:
Merge remote-tracking branch 'upstream/master' into opsmixin

Dr-Irv committed:
Merge remote-tracking branch 'upstream/master' into opsmixin

Dr-Irv committed:
Fix up merge issue

Dr-Irv committed:
Merge remote-tracking branch 'upstream/master' into opsmixin

Dr-Irv committed:
Remove changes related to get() bug fix

Dr-Irv commented:
@jreback following up on the conversation at the pandas dev meeting, I have removed the code that is in #21260 (since this PR no longer needs it), so I think this one here is good to go.

jorisvandenbossche commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -12,6 +12,45 @@ New features
 
 .. _whatsnew_0240.enhancements.other:
```
this label belongs to the title below your added content

Dr-Irv replied:
fixed

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +615,125 @@ def _ndarray_values(self):
+
+        Parameters
+        ----------
+        op: function
```
space after "op" and before the colon (numpy docstring standard, long version: http://pandas.pydata.org/pandas-docs/stable/contributing_docstring.html) 

and the same below

Dr-Irv replied:
fixed

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +615,125 @@ def _ndarray_values(self):
+
+        def _binop(self, other):
+            def convert_values(param):
+                if isinstance(param, ExtensionArray) or is_list_like(param):
```
Should we make this `is_list_like(param)` more strict as `is_array_like` ? 

For example, if you create ExtensionArray of sets, and do an operation where the right value is a single set, the current code will not work.

jorisvandenbossche replied:
Or similar with dicts (for which we already have a dummy implementation in the tests)

(I am also fine with leaving this for later, as there are other places where we have problems with iterable scalar elements)

Dr-Irv replied:
@jorisvandenbossche I think I want to leave it for now.  Because you'd like to be able to do an operation such as `EABackedSeries + list(objects)` and using `is_array_type` means you have to have a dtype.

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +615,125 @@ def _ndarray_values(self):
+                try:
+                    res = self._from_sequence(res)
+                except TypeError:
+                    pass
```
if this fails, I think we should still convert it to an array instead of keeping it as a list?

Or does that happen on another level? 
(anyway, also for usability with the ExtensionArray itself, I think coercing it to an array makes more sense)

Dr-Irv replied:
If we convert to an array, then we could have a dtype problem. This allows the result to be of any type.

jorisvandenbossche replied:
> If we convert to an array, then we could have a dtype problem

It will be converted to an array anyhow, if not here, then at the level above when the `res` is passed to the series constructor. So personally I would already do the conversion here (we can use the same logic / code as what is done in the series constructor to coerce a passed list)

Dr-Irv replied:
@jorisvandenbossche But why repeat that logic?  If we leave it as a list, then the Series constructor will do the inference on the dtype.

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +615,125 @@ def _ndarray_values(self):
+
+            return res
+
+        op_name = ops._get_op_name(op, True)
```
can you use a parameter name instead of the positional argument ?

Dr-Irv replied:
I could do that, but not specifying the parameter is consistent with all the other usages of `_get_op_name`, so I think I should be consistent with code that is elsewhere.

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -610,3 +615,125 @@ def _ndarray_values(self):
+
+    @classmethod
+    def _create_comparison_method(cls, op):
+        return cls._create_method(op, False)
```
same here, `coerce_to_dtype=False`

Dr-Irv replied:
fixed

jorisvandenbossche commented on pandas/tests/extension/base/ops.py:
```diff
@@ -0,0 +1,89 @@
+
+    def test_compare_scalar(self, data, all_compare_operators):
+        op_name = all_compare_operators
+        self._compare_other(data, op_name, 0)
```
for the arithmetic tests above you convert `data` to `s = pd.Series(data)` to perform the test on. Should do that here as well? 
Also, should we compare with `data[0]` instead of `0` ?

jorisvandenbossche replied:
Ah, I see that here that is done in the `_compare_other`. However, I would move that here to a) be consistent and b) be more explicit by looking at the single test code that it are the Series methods we are testing.

Dr-Irv replied:
made both changes

jorisvandenbossche commented on pandas/tests/extension/base/ops.py:
```diff
@@ -0,0 +1,89 @@
+        else:
+
+            # array
+            getattr(data, op_name)(other) is NotImplementedError
```
Why does this return NotImplementedError?

Dr-Irv replied:
Copied code from @jreback .  Changed now to have an assert and that it is NotImplemented

jorisvandenbossche commented on pandas/tests/extension/base/ops.py:
```diff
@@ -0,0 +1,89 @@
+    def _compare_other(self, data, op_name, other):
+        s = pd.Series(data)
+
+        if op_name == '__eq__':
```
I think you should do here something similar as in the Arithmetic tests to convert the `__eq__` to `operator.eq` (so to not test `s.__eq__(other)` directly but to test `s == other`

Dr-Irv replied:
I did a mix.  We want to test that by default the attributes of the op_name are not there, but we still want to call the operators directly and see that they raise `TypeError` (except for eq and ne which are defined)

jorisvandenbossche commented on pandas/tests/extension/decimal/test_decimal.py:
```diff
@@ -191,3 +191,63 @@ def test_dataframe_constructor_with_different_dtype_raises():
+
+    def check_opname(self, s, op_name, other, exc=None):
+        super(TestArithmeticOps, self).check_opname(s, op_name,
+                                                    other, exc=None)
```
Is this needed? It seems you are not changing anything in the implementation

Dr-Irv replied:
Yes, this is needed because the super `check_opname` has a default of `exc=NotImplementedError`.  So this overrides that default.

jorisvandenbossche commented on pandas/util/testing.py:
```diff
@@ -1247,6 +1248,10 @@ def assert_series_equal(left, right, check_dtype=True,
 
+    elif (is_extension_array_dtype(left) and not is_categorical_dtype(left) and
+          is_extension_array_dtype(right) and not is_categorical_dtype(right)):
+        return assert_extension_array_equal(left.values, right.values)
```
I think we should only do this if `check_dtype=True` ?

Dr-Irv replied:
I'm not sure I follow your logic for this.  Before, we had no code that really compared two EA's for testing. However, that code doesn't work correctly if they are categorical dtypes.  The `check_dtype` is a separate check, so if `check_dtype` was False, you'd still do this test the same way.

Dr-Irv committed:
changes per joris comments

jreback committed:
Merge branch 'master' into PR_TOOL_MERGE_PR_21261

jreback committed:
doc & clean

jreback committed:
moar doc

jreback commented:
thanks!

jorisvandenbossche commented:
@jreback there were still some unresolved comments. @Dr-Irv can you do a follow-up PR for the case where an EA is the right hand side of operation?

jorisvandenbossche commented:
2018-06-29 15:51 GMT+02:00 Dr. Irv <notifications@github.com>:

> *@Dr-Irv* commented on this pull request.
> ------------------------------
>
> In pandas/core/arrays/base.py
> <https://github.com/pandas-dev/pandas/pull/21261#discussion_r199165171>:
>
> > +                    ovalues = param
> +                else:  # Assume its an object
> +                    ovalues = [param] * len(self)
> +                return ovalues
> +            lvalues = self
> +            rvalues = convert_values(other)
> +
> +            # If the operator is not defined for the underlying objects,
> +            # a TypeError should be raised
> +            res = [op(a, b) for (a, b) in zip(lvalues, rvalues)]
> +
> +            if coerce_to_dtype:
> +                try:
> +                    res = self._from_sequence(res)
> +                except TypeError:
> +                    pass
>
> @jorisvandenbossche <https://github.com/jorisvandenbossche> But why
> repeat that logic? If we leave it as a list, then the Series constructor
> will do the inference on the dtype.
>
>
> Because IMO the ExtensionArray itself should also do the sensible thing
(eg sometimes you want to act on the values, eg if you don't want
alignment). And returning a list is then not consistent and not useful.
Note that it's not fully repeating the logic, it's only reusing the
underlying functionality (in principle it should be a single helper
function, but would need to look that up), and no inference needs to happen
in the Series constructor.
